### PR TITLE
Kbs automata integration

### PR DESCRIPTION
**Pull request by mikolajpabiszczak**
So this is a very crude integration of Index Automata with Knuth-Bendix procedure (2nd version) - the automaton is updated (i.e. `makeindexautomaton!` is run) every time the `RewritingSystem` is updated. Other changes:

- `Automata` are slightly redesigned (new field `stateslengths` which was moved from the `makeindexautomaton`);
- updated `makeindexautomaton`;
- I think that at some point an error got into `rewrite_from_left` which uses `Automaton` - which I corrected.

I will also add more tests: publishing it so that you can have a look.

And of course a quick benchmark:
```julia
using KnuthBendix
using BenchmarkTools
using Random
using Test
include("test/rws_examples.jl")


let R = RWS_Example_237_abaB(4), seed=1
    @btime KnuthBendix.knuthbendix2($R)
    @btime KnuthBendix.knuthbendix2automaton($R)
end;
```

which gives:
```julia
  30.099 ms (22105 allocations: 1.76 MiB)
  11.942 ms (145458 allocations: 11.56 MiB)
```

So even though we update the automaton every time the `RewritingSystem` changes we are faster than normal `kbs2`.

### CODE REVIEW

File: src/automata.jl [hunk 1]
```diff
@@ -294,7 +297,7 @@ function makeindexautomaton(rws::RewritingSystem, abt::Alphabet=alphabet(orderin
             end
         end
         i += 1
-        indcs = findall(isequal(i), Σᵢ)
+        indcs = findall(isequal(i), stateslengths(a))
```
Comment (kalmarek) [inline src/automata.jl:308]:
```julia
resize!(idcs, 0)
for (k,len) in enumerate(stateslengths(a))
    len == i && push!(idcs, k)
end
```

Comment (kalmarek) [reply src/automata.jl:308]: also above: I think that `τ = last(walk(a, @view(name(σ)[2:end])))` should cut allocations as well

File: src/automata.jl [hunk 2]
```diff
@@ -307,14 +335,13 @@ to `v`. For standard rewriting `v` should be empty.
...
+    resize!(past_states, length(w) + 1)
     state = initialstate(a)
     past_states[1] = state
-    initial_length = length(v)
 
     while !isone(w)
         x = popfirst!(w)
-        k = length(v) - initial_length + 1
+        k = length(v) + 1
         state = outedges(past_states[k])[x]
```
Comment (kalmarek) [inline src/automata.jl:345]: why this change? If `v` is non-empty (we're just appending the rewrite procedure to existing `v`) this will fail:
`length(w)` might be small compared to `length(v)` and we get out of bounds

**Review by kalmarek — CHANGES_REQUESTED** [review]

File: src/automata.jl (hunk 2 above)
Comment (mikolajpabiszczak) [reply src/automata.jl:345]: Ahh, yes that's true. I forgot for a moment about the way interface to rewriting is designed. `k = length(v) - initial_length + 1` will be ok. Sorry for that.

File: src/automata.jl (hunk 1 above)
Comment (mikolajpabiszczak) [reply src/automata.jl]: This suggestion is good, but it has to be applied in line 308 (in line 300 we first declare `indcs`).

**Comment by mikolajpabiszczak** [comment]
And the same benchmark gives now:

```julia
  30.129 ms (22105 allocations: 1.76 MiB)
  11.001 ms (116624 allocations: 9.03 MiB)
```

so there is some gain.

**Comment by mikolajpabiszczak** [comment]
Wrt. our conversation today - another benchmark:
```julia
using KnuthBendix
using BenchmarkTools
using Random
using Test
include("test/rws_examples.jl")


let R = RWS_Example_237_abaB(6), seed=1
    @btime KnuthBendix.knuthbendix2($R, maxrules=300)
    @btime KnuthBendix.knuthbendix2automaton($R, maxrules=300)
end;
```
and results:
```julia
  2.536 s (217613 allocations: 21.17 MiB)
  281.123 ms (1330914 allocations: 111.83 MiB)
```

Which I think is interesting.

**Comment by kalmarek** [comment]
very nice!

**Comment by mikolajpabiszczak** [comment]
Also tried to test `RWS_Example_237_abaB(8)` - I think at the moment we can call it non-reachable: `maxrules=1000` is not enough.

**Comment by kalmarek** [comment]
and how long does it take for index automata (with 1000 rules?)

**Comment by mikolajpabiszczak** [comment]
```julia
julia> R = RWS_Example_237_abaB(8)
julia> @time KnuthBendix.knuthbendix2automaton(R, maxrules=1000);
┌ Warning: Maximum number of rules (1000) in the RewritingSystem reached.
│                 You may retry with `maxrules` kwarg set to higher value.
└ @ KnuthBendix ~/Documents/GitHub/KnuthBendix.jl/src/automata_kbs2.jl:83
250.637719 seconds (70.67 M allocations: 10.361 GiB, 0.66% gc time)
```

**Review by kalmarek — APPROVED** [review]
